File: frontend/src/collections/cognitive.ts
import { createStore } from "@tanstack/react-store";

/*
CognitiveReading mirrors the backend types.Cognition payload exported on each
Thesis tick. It carries the sensory prefix tree, beam paths, and basin posteriors
that Cortex renders from thesis.cognition frames.
*/
export type CognitiveReading = {
	scope: string;
	sequence: string;
	regimePrefix: string;
	regimeCohort: number;
	ambiguous: boolean;
	sideline: boolean;
	entropyBits: number;
	entropyThreshold: number;
	classConfidence: number;
	contrastEvidence: number;
	lookaheadScore: number;
	lookaheadPaths: number;
	winnerClass: string;
	prewarmPaths?: number | null;
	prewarmScore?: number | null;
	updatedAt: number;
	beamWidth?: number;
	maxHops?: number;
	nodeCount?: number;
	branches?: CognitiveBranch[];
	beams?: CognitiveBeam[];
	classes?: CognitiveClass[];
	remFrom?: string;
	remThrough?: string;
	remReplays?: number;
};

export type CognitiveBranch = {
	id: number;
	parentId: number;
	token: string;
	prefix: string;
	depth: number;
	probability: number;
	count: number;
};

export type CognitiveBeam = {
	sequence: string;
	score: number;
};

export type CognitiveClass = {
	name: string;
	probability: number;
};

type CognitionFrame = {
	symbol?: string;
	at?: string;
	sequence?: string;
	regimePrefix?: string;
	winner?: string;
	ready?: boolean;
	confidence?: number;
	contrast?: number;
	entropyBits?: number;
	entropyThreshold?: number;
	ambiguous?: boolean;
	cohort?: number;
	lookaheadScore?: number;
	lookaheadPaths?: number;
	beamWidth?: number;
	maxHops?: number;
	nodeCount?: number;
	branches?: CognitiveBranch[];
	beams?: CognitiveBeam[];
	classes?: CognitiveClass[];
	remFrom?: string;
	remThrough?: string;
	remReplays?: number;
};

const finite = (value: unknown): number | null =>
	typeof value === "number" && Number.isFinite(value) ? value : null;

const stringField = (value: unknown, fallback = ""): string =>
	typeof value === "string" ? value : fallback;

/*
cognitiveReadingFromFrame maps one backend Cognition record into the Cortex store
shape without dropping tree visualization fields.
*/
export const cognitiveReadingFromFrame = (
	frame: CognitionFrame,
): CognitiveReading | null => {
	const scope = stringField(frame.symbol);

	if (scope === "") {
		return null;
	}

	const updatedAt = Date.parse(stringField(frame.at));

	return {
		scope,
		sequence: stringField(frame.sequence),
		regimePrefix: stringField(frame.regimePrefix),
		regimeCohort: finite(frame.cohort) ?? 0,
		ambiguous: frame.ambiguous === true,
		sideline: frame.ready === false,
		entropyBits: finite(frame.entropyBits) ?? 0,
		entropyThreshold: finite(frame.entropyThreshold) ?? 0,
		classConfidence: finite(frame.confidence) ?? 0,
		contrastEvidence: finite(frame.contrast) ?? 0,
		lookaheadScore: finite(frame.lookaheadScore) ?? 0,
		lookaheadPaths: finite(frame.lookaheadPaths) ?? 0,
		winnerClass: stringField(frame.winner),
		updatedAt: Number.isFinite(updatedAt) ? updatedAt : Date.now(),
		beamWidth: finite(frame.beamWidth) ?? undefined,
		maxHops: finite(frame.maxHops) ?? undefined,
		nodeCount: finite(frame.nodeCount) ?? undefined,
		branches: frame.branches,
		beams: frame.beams,
		classes: frame.classes,
		remFrom: frame.remFrom,
		remThrough: frame.remThrough,
		remReplays: finite(frame.remReplays) ?? undefined,
	};
};

const readingsFromFrame = (
	frame: unknown,
): Record<string, CognitiveReading> | null => {
	if (Array.isArray(frame)) {
		const readings: Record<string, CognitiveReading> = {};

		for (const entry of frame) {
			if (entry === null || typeof entry !== "object") {
				continue;
			}

			const reading = cognitiveReadingFromFrame(entry as CognitionFrame);

			if (reading === null) {
				continue;
			}

			readings[reading.scope] = reading;
		}

		return Object.keys(readings).length === 0 ? null : readings;
	}

	if (frame === null || typeof frame !== "object") {
		return null;
	}

	const record = frame as Record<string, unknown>;
	const nestedReadings = record.readings;

	if (nestedReadings !== undefined && nestedReadings !== null) {
		return readingsFromFrame(nestedReadings);
	}

	const directReading = cognitiveReadingFromFrame(record as CognitionFrame);

	if (directReading === null) {
		return null;
	}

	return { [directReading.scope]: directReading };
};

/*
cognitiveStore holds the latest per-symbol cognitive readings broadcast on each
Thesis tick under cognition. The Cortex surface renders the DMT sequence tree,
entropy gate, and posterior from these.
*/
export const cognitiveStore = createStore(
	{
		readings: {} as Record<string, CognitiveReading>,
		selectedScope: null as string | null,
	},
	({ setState }) => ({
		updateFrame: (frame: unknown) =>
			setState((prev) => {
				const readings = readingsFromFrame(frame);

				if (readings === null) {
					return prev;
				}

				return {
					...prev,
					readings: {
						...prev.readings,
						...readings,
					},
				};
			}),
		selectScope: (selectedScope: string) =>
			setState((prev) => ({ ...prev, selectedScope })),
	}),
);

/*
cognitiveScopes lists the symbols that currently have a cognitive reading in a
stable order.
*/
export const cognitiveScopes = (
	readings: Record<string, CognitiveReading>,
): string[] => Object.keys(readings).sort();

File: market/cognitive.go
package market

import "github.com/theapemachine/symm/logic"

/*
CognitiveReading is the per-symbol cognitive summary the Cortex surface renders.
The DMT evaluator builds it, but the UI and trader consume it as plain state.
*/
type CognitiveReading struct {
	Scope            string            `json:"scope"`
	Sequence         string            `json:"sequence"`
	RegimePrefix     string            `json:"regimePrefix"`
	RegimeCohort     int               `json:"regimeCohort"`
	Ambiguous        bool              `json:"ambiguous"`
	Sideline         bool              `json:"sideline"`
	EntropyBits      float64           `json:"entropyBits"`
	EntropyThreshold float64           `json:"entropyThreshold"`
	Surprisal        float64           `json:"surprisal"`
	Surprise         float64           `json:"surprise"`
	ClassConfidence  float64           `json:"classConfidence"`
	ContrastEvidence float64           `json:"contrastEvidence"`
	LookaheadScore   float64           `json:"lookaheadScore"`
	LookaheadPaths   int               `json:"lookaheadPaths"`
	WinnerClass      string            `json:"winnerClass"`
	UpdatedAt        int64             `json:"updatedAt"`
	BeamWidth        int               `json:"beamWidth"`
	MaxHops          int               `json:"maxHops"`
	NodeCount        int               `json:"nodeCount"`
	Branches         []CognitiveBranch `json:"branches,omitempty"`
	Beams            []CognitiveBeam   `json:"beams,omitempty"`
	Classes          []CognitiveClass  `json:"classes,omitempty"`
}

type CognitiveBranch struct {
	ID          int     `json:"id"`
	ParentID    int     `json:"parentId"`
	Token       string  `json:"token"`
	Prefix      string  `json:"prefix"`
	Depth       int     `json:"depth"`
	Probability float64 `json:"probability"`
	Count       uint64  `json:"count"`
}

type CognitiveBeam struct {
	Sequence string  `json:"sequence"`
	Score    float64 `json:"score"`
}

type CognitiveClass struct {
	Name        string  `json:"name"`
	Probability float64 `json:"probability"`
}

/*
ApplyCognitiveReadings stamps live measurements with the cognitive state read
from the backend evaluator.
*/
func ApplyCognitiveReadings(
	measurements []*logic.Measurement,
	readings map[string]CognitiveReading,
) {
	if len(measurements) == 0 || len(readings) == 0 {
		return
	}

	for _, measurement := range measurements {
		if measurement == nil {
			continue
		}

		if measurement.Symbol == "" {
			continue
		}

		reading, ok := readings[measurement.Symbol]

		if !ok {
			continue
		}

		surprise := reading.Surprise

		if surprise == 0 && reading.EntropyThreshold > 0 {
			surprise = reading.EntropyBits / reading.EntropyThreshold
		}

		if measurement.Metrics == nil {
			measurement.Metrics = map[string]float64{}
		}

		measurement.Metrics["surprisal"] = reading.Surprisal
		measurement.Surprise = surprise
		measurement.Metrics["surpriseThreshold"] = reading.EntropyThreshold
		measurement.Metrics["cognitiveClassConfidence"] = reading.ClassConfidence
		measurement.Status = cognitiveMeasurementStatus(measurement, reading)
	}
}

func cognitiveMeasurementStatus(
	measurement *logic.Measurement,
	reading CognitiveReading,
) string {
	confidence := measurement.Confidence
	strength := measurement.Strength

	if confidence <= 0 || strength <= 0 {
		return "standby"
	}

	if reading.Sideline || reading.Ambiguous {
		return "ambiguous"
	}

	if reading.ClassConfidence <= 0 {
		return "calibrating"
	}

	if measurement.Source == logic.SourceCausal && !measurement.CounterfactualReady {
		return "calibrating"
	}

	return "measured"
}
